### PR TITLE
TST: add hypothesis-based tests

### LINKED ISSUE #17978: Investigate using Hypothesis for some tests
Hypothesis recently added support for generating pandas objects: http://hypothesis.readthedocs.io/en/latest/numpy.html#pandas

This could be useful for testing some of the higher-level methods like `melt`, `merge`, etc.

Are there any objections to adding some hypothesis-based tests to our test suite? Is anyone familiar with hypothesis and interested in putting together examples of how we could use it?

### PR DESCRIPTION
**Pull request by sushobhit27**
Addition of "hypothesis usage" in test cases of
tests/reshape/test_util.py as kind of POC.

- [x] closes #17978
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
> Could you please check how it (hypothesis module) can be available, so that tests can be run.

You'll need to add hypothesis to one (or all) of the `ci/*.run` files. Also add it to `ci/environment-dev.yml`.

**Comment by sushobhit27** [comment]
@TomAugspurger Now all test have passed. Can you please check what else is required at my end.

**Review by jreback — CHANGES_REQUESTED** [review]
will have to look
this is adding lots of code in a very specific file
need a more general way to do this

**Comment by jreback** [comment]
test are failing in this

File: pandas/tests/reshape/test_util.py [hunk 1]
```diff
@@ -4,31 +4,101 @@
...
+    elif elem_type == float:
+        strategy = st.floats()
+    elif elem_type == str:
+        strategy = st.text(string.ascii_letters, max_size=10)
+    return strategy
+
+
+@st.composite
+def get_seq(draw, types, mixed=False, min_size=None, max_size=None, transform_func=None):
+    """helper function to generate strategy for creating lists. parameters define the nature of to be generated list.
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_util.py]: These lines are too long (check the log output or run flake8 locally)

Comment (TomAugspurger) [reply pandas/tests/reshape/test_util.py]: Also, we use the numpy docstring standard: http://pandas-docs.github.io/pandas-docs-travis/contributing_docstring.html

File: pandas/tests/reshape/test_util.py [hunk 2]
```diff
@@ -4,31 +4,101 @@
...
+
         tm.assert_numpy_array_equal(result1, expected1)
         tm.assert_numpy_array_equal(result2, expected2)
 
+    @settings(max_examples=NO_OF_EXAMPLES_PER_TEST_CASE)
     def test_datetimeindex(self):
         # regression test for GitHub issue #6439
         # make sure that the ordering on datetimeindex is consistent
-        x = date_range('2000-01-01', periods=2)
+        d = st.dates(min_value=date(1900, 1, 1), max_value=date(2100, 1, 1)).example()
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_util.py]: Why generate them here instead of using `@given`?

**Review by TomAugspurger — COMMENTED** [review]
Most contributors -- and maintainers :) -- will be unfamiliar with hypothesis. Could you add a section to the contributing documentation, along with links to the hypothesis docs? Mainly I'd be interested in hearing about what tests makes sense for hypothesis, and perhaps a small examples using it.

File: pandas/tests/reshape/test_util.py (hunk 2 above)
Comment (sushobhit27) [reply pandas/tests/reshape/test_util.py]: no specific reason but copy pasted after testing on console, as example() function can be easily used to check generated value  :) and ya it can be moved to "given" decorator.

**Comment by TomAugspurger** [comment]
To be clear, I wasn't suggesting a change. I haven't used hypothesis much.
It'd be good to establish best practices since this will be new to
pandas contributors.

**Comment by sushobhit27** [comment]
@TomAugspurger, 
Main purpose of using hypothesis or property based testing is to just write test cases based on the property/specification of the functionality being tested and let the framework generate all the different and random test cases for it. 
This will initially make the test cases fail but due to which we are able to find edge cases and then keep on refining boundary for input data of our test cases. e.g in below example initially test case failed when x was a list with empty string, thus in a way helped me to figure out that this is a case for test_empty function (due to which I adeed assume(len(x) != 0) statement). 

    def test_simple(self):
        x, y = list('ABC'), [1, 22]
        result1, result2 = cartesian_product([x, y])
        expected1 = np.array(['A', 'A', 'B', 'B', 'C', 'C'])
        expected2 = np.array([1, 22, 1, 22, 1, 22])
        tm.assert_numpy_array_equal(result1, expected1)
        tm.assert_numpy_array_equal(result2, expected2)
		

	@settings(max_examples=NO_OF_EXAMPLES_PER_TEST_CASE)
    @given(get_seq((str,), False, 1, 1),
           get_seq((int,), False, 1, 2))
    def test_simple(self, x, y):
        x = list(x[0])
        # non-empty test case is handled in test_empty, therefore ignore it here
        assume(len(x) != 0)
        result1, result2 = cartesian_product([x, y])
        expected1 = np.array([item1 for item1 in x for item2 in y])
        expected2 = np.array([item2 for item1 in x for item2 in y])

        tm.assert_numpy_array_equal(result1, expected1)
        tm.assert_numpy_array_equal(result2, expected2)
		

In my personal experience, I have found it to be most helpful in reducing huge parametized test cases where many examples are overlapping or can be tested by a single hypothesis statement.
		
In very layman terms, take another example of testing python's sum function using hypothesis by just a single statement.
from hypothesis import strategies as st
from hypothesis import given

@given(st.lists(st.integers()))
def test_sum(seq):
    total = 0
    for item in seq:
        total += item
    assert sum(seq) == total

While, IMO in case of example based testing at least 3-4 example would be required in terms of parametize statement.

Below are some links for property bases testing usage:
https://hypothesis.readthedocs.io/en/latest/quickstart.html
http://blog.jessitron.com/2013/04/property-based-testing-what-is-it.html
https://hypothesis.works/articles/what-is-property-based-testing/

**Comment by TomAugspurger** [comment]
Great, could you summarize appropriately for a new subsection in
contributing.rst? :)

**Comment by sushobhit27** [comment]
@TomAugspurger sure, I will do the same in a day or two in addition to incorporating other review comments.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by sushobhit27** [comment]
@jreback , @TomAugspurger 
I know its going on for too long, but can you please have a look at latest changes.
Travis is failing due to other test case, TestClipboard.test_round_trip_valid_encodings to be precise.

File: ci/requirements-2.7_LOCALE.run [hunk 3]
```diff
@@ -10,3 +10,4 @@ matplotlib=1.4.3
 sqlalchemy=0.8.1
 lxml
 scipy
+hypothesis>=3.46.0
```
[the remark this answers is not in the document]
Comment (jreback) [reply ci/requirements-2.7_LOCALE.run]: I put them above

``ci/install_travis_travis``, ``ci/install_circle`` and ``appveyor.yaml``

File: doc/source/contributing.rst [hunk 4]
```diff
@@ -775,6 +775,80 @@ Tests that we have ``parametrized`` are now accessible via the test name, for ex
    test_cool_feature.py::test_dtypes[int8] PASSED
    test_cool_feature.py::test_series[int8] PASSED
 
+Using ``hypothesis``
+~~~~~~~~~~~~~~~~~~~~
+With the transition to pytest, things have become easier for testing by having reduced boilerplate for test cases and also by utilizing pytest's features like parametizing, skipping and marking test cases.
```
Comment (jreback) [inline doc/source/contributing.rst]: remove this 'transition to pytest

File: doc/source/contributing.rst [hunk 5]
```diff
@@ -775,6 +775,78 @@ Tests that we have ``parametrized`` are now accessible via the test name, for ex
...
+    @given(st.lists(st.integers()))
+    def test_sum(seq):
+        total = 0
+        for item in seq:
+            total += item
+        assert sum(seq) == total
+
+
+output of test cases:
+
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/contributing.rst]: the context, make this whole section shorter

File: pandas/util/_hypothesis.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import string
```
Comment (jreback) [inline pandas/util/_hypothesis.py]: add a doc-string to this module

File: pandas/util/_hypothesis.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+from hypothesis import (given,
+                        settings,
+                        assume,
+                        strategies as st,
+                        )
+
+
+def get_elements(elem_type):
+    strategy = st.nothing()
```
Comment (jreback) [inline pandas/util/_hypothesis.py]: add doc-strings to each

File: pandas/util/_hypothesis.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    Returns
+    -------
+    hypothesis lists strategy.
+
+    Examples
+    --------
+    seq_strategy = get_seq((int, str, bool),
+                            mixed=True, min_size=1, max_size=5)
+    seq_strategy.example()
+    Out[12]: ['lkYMSn', -2501, 35, 'J']
```
Comment (jreback) [inline pandas/util/_hypothesis.py]: separate examples with a blank line, use ``>>>`` as this should be a proper doc-string

File: pandas/tests/reshape/test_util.py [hunk 9]
```diff
@@ -4,31 +4,60 @@
 import pandas.util.testing as tm
 from pandas.core.reshape.util import cartesian_product
 
+import string
+from datetime import date
+from dateutil import relativedelta
+
+from pandas.util._hypothesis import (st,
```
Comment (jreback) [inline pandas/tests/reshape/test_util.py]: make this simpler.

``from pandas.util import _hypothesis as hp``

File: pandas/tests/reshape/test_util.py [hunk 10]
```diff
@@ -4,31 +4,60 @@
...
+                                     given,
+                                     settings,
+                                     get_seq,
+                                     assume)
+
+
+NO_OF_EXAMPLES_PER_TEST_CASE = 20
```
Comment (jreback) [inline pandas/tests/reshape/test_util.py]: hard-code this for now in the examples. have to see how this behaves.

File: pandas/tests/reshape/test_util.py [hunk 11]
```diff
@@ -4,31 +4,60 @@
...
+
 
 class TestCartesianProduct(object):
 
-    def test_simple(self):
-        x, y = list('ABC'), [1, 22]
+    @settings(max_examples=NO_OF_EXAMPLES_PER_TEST_CASE)
+    @given(get_seq((str,), False, 1, 1),
```
Comment (jreback) [inline pandas/tests/reshape/test_util.py]: is there any way to name these things for ``given``, pretty non-intuitive here

File: pandas/tests/reshape/test_util.py [hunk 12]
```diff
@@ -4,31 +4,60 @@
...
+           get_seq((int,), False, 1, 2))
+    def test_simple(self, x, y):
+        x = list(x[0])
```
Comment (jreback) [inline pandas/tests/reshape/test_util.py]: why is this is a list?

File: pandas/util/_hypothesis.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
+import string
+from hypothesis import (given,
+                        settings,
+                        assume,
+                        strategies as st,
```
Comment (jreback) [inline pandas/util/_hypothesis.py]: can you add some direct documentation / examples in this file

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_util.py (hunk 11 above)
Comment (sushobhit27) [reply pandas/tests/reshape/test_util.py]: We are already giving them a name in some cases, as arguments in test case function. e.g
@given(st.lists(st.nothing()),
           get_seq((int,), False, min_size=1, max_size=10),
           get_seq((str,), False, min_size=1, max_size=10))
    def test_empty(self, **empty_list, list_of_int, list_of_str**):

local function can be added like below:
def get_str_list_with_single_element():
    return get_seq((str,), False, 1, 1)
but that would be too cumbersome. Instead,  for each argument in given decorator, having comments explaining each returned strategy can be more suitable.

**Comment by sushobhit27** [comment]
@jreback, @TomAugspurger 
Was waiting for feedback, trying luck again :)

**Comment by jreback** [comment]
@sushobhit27 I indicated where you need to update these files in ``pandas/ci/install_circle``, ``pandas/ci/appeveyor.yml`` and ``pandas/ci/install_travis.sh``

**Comment by sushobhit27** [comment]
@jreback  I have tried to do changes as per your comments and now have removed hypothesis dependency from all *.run, except from the 3 files which you mentioned, although I am not sure if I did it correctly.
Let me know if it is still wrong.

File: ci/environment-dev.yaml [hunk 14]
```diff
@@ -13,3 +13,4 @@ dependencies:
   - pytz
   - setuptools>=3.3
   - sphinx
+  - hypothesis>=3.46.0
```
Comment (jreback) [inline ci/environment-dev.yaml]: need a new-line

File: ci/install_circle.sh [hunk 15]
```diff
@@ -65,6 +65,7 @@ fi
 echo "[create env: ${REQ_BUILD}]"
 time conda create -n pandas -q --file=${REQ_BUILD} || exit 1
 time conda install -n pandas pytest>=3.1.0 || exit 1
```
Comment (jreback) [inline ci/install_circle.sh]: you can just add it on the previous line

File: ci/install_travis.sh [hunk 16]
```diff
@@ -104,6 +104,7 @@ if [ -e ${REQ} ]; then
 fi
 
 time conda install -n pandas pytest>=3.1.0
```
Comment (jreback) [inline ci/install_travis.sh]: same

File: ci/requirements-2.7_COMPAT.run [hunk 17]
```diff
@@ -11,4 +11,4 @@ psycopg2
 pymysql=0.6.0
 sqlalchemy=0.7.8
 xlsxwriter=0.5.2
-jinja2=2.8
```
Comment (jreback) [inline ci/requirements-2.7_COMPAT.run]: revert these, they shouldn't have any changes

File: doc/source/contributing.rst [hunk 18]
```diff
@@ -775,6 +775,46 @@ Tests that we have ``parametrized`` are now accessible via the test name, for ex
    test_cool_feature.py::test_dtypes[int8] PASSED
    test_cool_feature.py::test_series[int8] PASSED
 
+Using ``hypothesis``
```
Comment (jreback) [inline doc/source/contributing.rst]: add a ref tag here

File: doc/source/contributing.rst [hunk 19]
```diff
@@ -775,6 +775,46 @@ Tests that we have ``parametrized`` are now accessible via the test name, for ex
...
+~~~~~~~~~~~~~~~~~~~~
+With the usage of pytest, things have become easier for testing by having reduced boilerplate for test cases and also by utilizing pytest's features like parametizing, skipping and marking test cases.
```
Comment (jreback) [inline doc/source/contributing.rst]: use double-backticks around pytest

File: doc/source/contributing.rst [hunk 20]
```diff
@@ -775,6 +775,46 @@ Tests that we have ``parametrized`` are now accessible via the test name, for ex
...
+
+However, one has to still come up with input data examples which can be tested against the functionality. There is always a possibility to skip testing an example which could have failed the test case.
+
+Hypothesis is a python package which helps in overcoming this issue by generating the input data based on some set of specifications provided by the user.
```
Comment (jreback) [inline doc/source/contributing.rst]: link to the package

File: pandas/tests/reshape/test_util.py [hunk 21]
```diff
@@ -4,31 +4,52 @@
...
+from pandas.util import _hypothesis as hp
+
+NO_OF_EXAMPLES_PER_TEST_CASE = 20
+
 
 class TestCartesianProduct(object):
 
-    def test_simple(self):
-        x, y = list('ABC'), [1, 22]
+    @hp.settings(max_examples=20)
```
Comment (jreback) [inline pandas/tests/reshape/test_util.py]: so this is a lot of boilerplate here, this has to be simpler, we have thousands of tests (sure most cannot use this), but I think you would need to have this reduced to basically a 1-liner to have people use it. can you put some functions in _hypthesis to make this much more readable.

this comment for each of the additions here.

**Review by jreback — CHANGES_REQUESTED** [review]
does this causes multiple tests to be run (like parameterized)? are these deterministic? these for sure need to be as we don't want non-deterministic ones on the ci (IOW that cannot be reproduced locally). if they are seed based this would be fine.

File: pandas/tests/reshape/test_util.py (hunk 21 above)
Comment (sushobhit27) [reply pandas/tests/reshape/test_util.py]: """so this is a lot of boilerplate here, this has to be simpler, we have thousands of tests (sure most cannot use this), but I think you would need to have this reduced to basically a 1-liner to have people use it. can you put some functions in _hypthesis to make this much more readable."""

If you are talking about boilerplate, like below decorators, I don't think, it can be further reduced as just like, we are bound to have different parametrize for different test cases in pytest, the same issue is with below code. For some function it will be less, for other it can be more. 
may be when code evolves, more common code comes out to be refactored.

@hp.settings(max_examples=20)
    @hp.given(hp.st.lists(hp.st.text(string.ascii_letters, min_size=1, max_size=1),
                          min_size=1, max_size=3),
           hp.get_seq((int,), False, 1, 2))
    def test_simple(self, x, y):

File: ci/requirements-2.7_COMPAT.run (hunk 17 above)
Comment (sushobhit27) [reply ci/requirements-2.7_COMPAT.run]: I will rebase my branch.

**Comment by TomAugspurger** [comment]
Note: we made some changes to the CI that will cause some merge conflicts,
though hopefully the new setup is simpler.

You can add hypothesis next to `pytest` in the new
`ci/{appveyor,circle,travis}-*.yaml` files.

**Comment by sushobhit27** [comment]
@TomAugspurger I totally forgot about this PR, now I have rebased branch. You can have a look at it now, as all tests are now passing.

**Comment by jreback** [comment]
closing in favor of #22280
